### ISSUE TITLE
Loadout weapons being moved around when unnecessary or incorrectly

### ISSUE THREAD

**Issue by M3Rocket** [dasilva333/TowerGhostForDestiny#236]
Currently, I have a loadout with 6 weapons circled in red:
![image](https://cloud.githubusercontent.com/assets/3769123/7925794/0301593a-087e-11e5-8c2a-d474903556c4.png)

I want to apply the Loadout to the Titan 33 but the loadout comes back with these errors:
![image](https://cloud.githubusercontent.com/assets/3769123/7925806/237478a0-087e-11e5-9b43-03c1dc3019ab.png)

**Comment by M3Rocket** [comment]
I feel like the logic for moving weapons/loadouts could use some fleshing out and improvement. 

Example:
Trying to loadout a Guardian with full primary/secondary/heavy slots results in "...will not be moved because there's no space in primary/secondary/heavy." This despite there being at least two spaces in the Vault or another character.

The logical operation should be based on the fact that you will need at minimum one open vault slot and one empty appropriate class weapon slot on a character:

1. Is there an empty weapon slot in the vault to move item?
1a. If there already is an empty weapon slot in vault or item to be moved is in the vault, continue to 2.
1b. If not, is there an empty slot anywhere else that a vault weapon can be moved to free up one slot in the vault? Move item to extra slot elsewhere to create at least one open slot in the vault.

2. Is there an empty slot on the destination for weapon?
2a. If yes, continue to 3.
2b. If not, move a weapon from destination to vault slot (if there are two or more vault slots). Alternatively, move weapon from destination to another character with an open slot if there is only one vault slot.

3. Move weapon to destination.

4. Repeat from step 1 until all weapons in loadout are addressed.


It seems to me that instead of swapping/moving items between characters, that the vault should be priority. Meaning if there is more than one weapon slot open in the vault, extra weapons on the destination that is full should always be moved and stored in the vault. It should not be swapped to a different character. Swapping weapons from the destination to another character by default (as it works right now) creates a lot of confusion. 

Also, when moving out weapons to clear out space in the destination, priority should be given to moving out identically named items. For instance I have two Fatebringers, one on Warlock and the other TItan. One is 365 attack and one is 300 attack. I want to "swap" the Fatebringers depending on which character I'm using. But a lot of times when I use a loadout to move the 365 Fatebringer, the destination character ends up with both Fatebringer's in their inventory. 

I think there should be some logic programmed in to "smartly" address situations above. So if both my source and destination have the same weapon but different attack values, then swap those weapons. Doesn't matter whether it's the Vault or other characters.

**Comment by Titus-s** [comment]
After HoW we all have all bucket and Vault almost filled up, and issues with no slots available will appear more frequently. 
I think that at start/refresh the application should check if there is a space in Vault on each category (Weapon / Armor / General). If there is no slots available it should inform (somehow) user that there is no way to transfer items. 
I thought about the way to inform user about it, maybe grey out item icons in characters bucket that belong to affected category? It should be visible on first sight, because it's blocking basic functionality of this app. 
For example: if there is no slot available in weapon bucket in Vault it should grey out all weapons icons in all characters. After click on greyed out icon - there should be information about no space in vault, and option to equip or store only.

**Comment by M3Rocket** [comment]
Technically, to move weapons and armor around, you need at the minimum one empty vault slot and one guardian weapon class specific slot. This does not apply to general materials and such because most of those items stack. But I agree--some sort of check and diagram or decision tree of what is possible is the only way to flesh out some of these "transfer" rules. 

I think DIM handles some things better because they are Guardian based loadouts--the rules are simpler. Loadouts in TGD is more akin to the iOS Loadouts app. Except for now, I think the Loadouts iOS app does a better job at moving weapons around.

**Comment by moeyjar** [comment]
This may be the same thing, but I'm also seeing some oddness with the way it's moving items around, even though there is space in both the vault, and on my characters.

In some cases, it will move without issue, but other times ,it wants to swap out a gun for another one, even though it's not trying to be equipped, or running low on space.  I was hoping that the loadouts would just move the guns instead of trying to do a swap, especially if i'm not trying to equip, or trying to move a gun without being any space left on the other character.

![loadout](https://cloud.githubusercontent.com/assets/12259797/8232243/6346173a-15a4-11e5-8fc9-6d28628de744.JPG)


![move](https://cloud.githubusercontent.com/assets/12259797/8232186/f28a45d4-15a3-11e5-833f-a9094d01cea4.JPG)
![inv](https://cloud.githubusercontent.com/assets/12259797/8232185/f2890bec-15a3-11e5-9d53-7b1747b22375.JPG)

**Comment by moeyjar** [comment]
It seems it may have something to do with the Exotics.  

- On hunter, I moved both Thorn and TLW to Vault
- I removed Vex from the Prison of Elders loadout

It now will move without any swapping

**Comment by M3Rocket** [comment]
Just another thought based on my experience. I have duplicates of certain weapons like Fatebringer and Gally that I don't need switched. So I usually leave them out of the Loadouts. Unfortunately, while switching other weapons around, often times, TGD will swap out the duplicates even when there is lots of space, and I end up with one guardian having two Fatebringers, etc. Meanwhile, I have to manually transfer them back to the other Guardian.

I could include the dupes in the loadout, but then I may end up with two Fatebringers, etc. on the destination Guardian since the logic for "swapping" weapons seem quite random to me. Or create a situation where there two similar stat Fatebringers are swapped needlessly between the source and destination, prolonging the loadout transfer.

Alternatively, I could create separate loadouts specific to each Guardian. Then it would work like DIM. But that makes it so troublesome when there's a change in the loadout--I then have to go to 3 different loadouts to make any changes.

**Comment by M3Rocket** [comment]
TL;DR There needs to be some logic when swapping weapons so that if the destination already has  same name weapon with different stats, those are the ones that need to be prioritized. So if the loadout contains a Fatebringer, check to see if the destination has a Fatebringer. If it does, then make sure these are swapped.

**Comment by rkingsla** [comment]
Hey all....  Was going to comment in 236, but we're trying to consolidate here.  So - here we are.  :)
I've a load out that I use nearly exclusively for Prison of Elders.  and, it loads a lot of stuff because I never quite know what I'm going to meet... And, I'm a hoarder.  I'm getting help on that...  Regardless, I run into the error "There are no item slots to transfer this item." or the even more fun "cannot move equipped weapon".

Suggestion to alleviate difficulty a the other issues are being worked out:
1. Tell me what Item could not be moved.  I'll self-resolve if needed. Sometimes it takes a bit of mucking around to figure out WHAT, as the WHY may not matter right now.
2. If an item could not be moved because it's equipped - tell me which item. I can only assume we know, because it's equipped and we tried to move it, so there is some reference back to a user displayable name, I'd assume.
And, I'll echo the @M3Rocket post from above. I'd LOVE to have an option to mark some items NEVER, EVER move unless I tell you to.  For example, I have an Ice Breaker and a Gjallarhorn for each toon. I don't want them moved.  I thought the answer MIGHT BE, well add them to the loadout.  Certainly we're going to see that it's in the special/heavy bucket of the target. Of course, I would be wrong... Well, it seems that the bungie.api being messed up so that some items seem to need XP when it really doesn't is causing this some issue and I think we can all agree that any character only needs on IB or Jelly to be effective at killing stuff.  :o)

**Comment by M3Rocket** [comment]
One "bug" I found is that if you equip a weapon in game, but in TGD it shows as unequipped (not refreshed), you'll get the "cannot move equipped weapon" message simply because TGD tries to move the item that is equipped. If it's equipped in TGD, then it swaps out a weapon before moving it. Simple solution would be to do a sectional refresh before making any moves within that weapon type.

But there are a LOT of scenarios that need to be mapped out. I'm not clear if @dasilva333 already has a master logic chart to how moving weapons work, but that's what we definitely need to figure out the best way to move weapons around.

**Comment by dasilva333** [comment]
Hey guys I revamped the logic for TGD in 2.9.2 so transfers should be handled a lot better now, I look forward to any feedback you all can provide as of this latest revision. I'll leave this ticket open for a bit and close it after I'm sure it's working as expected.

Thanks guys

**Comment by moeyjar** [comment]
This seems to work much better, thank you.  The only thing that would be nice (if it needs to spin as a new request, I can create it) but for those who have certain guns, lets say 3 gjallahorns, and you don't want them to be swapped, having a way to exclude it from being part of the swap process.  In my case, I don't have a ghorn in my loadout, as each character has 1.  In certain cases, one character will end up with 2, since it was involved in the move process.

The other thing that would be nice to exclude would be Decoherent Engrams, unless no other guns can be swapped in their place.  Reason being, is that they're typically disposable and sometimes disappear if you don't refresh the TGD app very often (I refresh mine everytime I use a loadout now).

Thanks again!

**Comment by dasilva333** [comment]
Thanks for the feedback @moeyjar you're request has already been posted in issue #100 I too have the same problem so I'll keep thinking about a good way to get that accomplished.

**Comment by purepvd** [comment]
Since this update I'm now seeing some unwanted behavior I can't explain well. I'm moving a loadout, receive the Transfer Confirm loadout summary, I click transfer, and for some reason I receive another Transfer Confirm pop-up asking me to confirm a single weapon transfer. I click Transfer again to get things moving (somewhat annoying), and the app moves that single weapon only - never moving the remainder of the loadout. Also odd, but I don't think consistent - in this image both weapons are in the loadout so they shouldn't need to be swapped in the first place.

![image](https://cloud.githubusercontent.com/assets/3674827/9306803/8ae06770-44c8-11e5-9d20-34f9e5c8fe6a.png)

**Comment by dasilva333** [comment]
I'm glad I didn't close this ticket I knew this issue was going to be posted sooner rather than later. Thanks @purepvd for bringing this up I too noticed it during my testing last night, to make matters worse clicking on Transfer completes that action but doesn't resume the loadouts sequence so it breaks the logic again. I'll be looking into why this is happening and coming up with a fix. If you could help me find a clear reproducible test case for getting that error to trigger that would be great.

Thanks again

**Comment by moeyjar** [comment]
I actually saw this same behavior when I was trying to move a gun to a full vault.  

Steps to reproduce.
1)  Fill up weapons to 36/36
2)  Try to move a gun to the vault, which should trigger a move sequence.  The dialog never seems to go away.

![transfer](https://cloud.githubusercontent.com/assets/12259797/9309597/21a0e76c-44e1-11e5-8701-f6985572100b.JPG)

**Comment by dasilva333** [comment]
@moeyjar thanks that's a great test I can reproduce that pretty easily expect it fixed in the next version thanks again

**Comment by edmiller47** [comment]
I started having the same problem as @purepvd and @moeyjar described today after updating to 2.9.4.
My loadouts generally have 8 inventory + 1 equipped weapon for each of the primary, special, and heavy slots. It pretty much is happening every time I try to transfer one of these loadouts.

I'd be happy to test your fix if you can tell me how to get it.

**Comment by dasilva333** [comment]
@edmiller47 can you describe the steps to reproduce the issue in 2.9.5? I can't seem to figure out how to get it to mess up anymore. Let me know thanks, as for going to an earlier build I'm afraid that's not possible sorry man

**Comment by edmiller47** [comment]
I'm happy to, will send you a note in the morning.

**Comment by edmiller47** [comment]
2.9.5 - running the Chromoe extension:


Here’s a view of my gear and weapons upon logging in. Look at the Titan –
one empty slot in each weapon bucket.

[image: Inline image 1]

Here’s what one of my loadouts looks like. Each loadout transfers 9 guns
per bucket (8 + 1 marked to be equipped).

[image: Inline image 2]

Applying the “Titan PoE” loadout to my Titan:

[image: Inline image 3]

Press “Transfer” here, and get the “completed successfully” popup, but
nothing else got transferred. The long “Transfer Confirm” list stays on the
screen until I press “Cancel”. Now my Primary bucket has all slots filled,
but the rest of the loadout didn’t get transferred. For example,  Helm of
St-14 didn’t get equipped.

[image: Inline image 4]

This happens with every one of my class-specific loadouts. If I apply the
same loadouts using 2.9.1 on my iPad, none of this stuff happens.
Everything gets transferred and equipped as specified in the loadout. Hope
this helps. Let me know if you need other information.

**Comment by dasilva333** [comment]
@edmiller47 the inline images are not visible or included in your post. I'll do further testing on my own to verify the claims. Thanks man

**Comment by edmiller47** [comment]
Sorry, updated the post - you should be able to view them now

**Comment by dasilva333** [comment]
@edmiller47 can you confirm if you still have any problems as of 2.9.6? Thanks

**Comment by purepvd** [comment]
I can confirm no issues thus far.

**Comment by edmiller47** [comment]
2.9.6 is working with no issues. I think I beat on it as hard as I could this morning and couldn't get it to fail, even when I filled up some of the weapon buckets. Awesome...

**Comment by dasilva333** [comment]
I'll leave it open for a bit longer and then create a new issue for the problem I noticed last night. Basically if you're playing a game and u want to xfer 10 items to that bucket there is one item that wont be able to get transferred because it can't be unequipped while you're playing so I'll have to add some logic to that. Thanks for the feedback
